Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -215,62 +184,71 @@ extension SwiftLanguageServer {
     }
     buildSettingsByFile[url] = newSettings
 
-    // Close and re-open the document internally to inform sourcekitd to
-    // update the settings. At the moment there's no better way to do this.
-    let closeReq = SKRequestDictionary(sourcekitd: sourcekitd)
-    closeReq[keys.request] = requests.editor_close
-    closeReq[keys.name] = url.path
-    _ = self.sourcekitd.sendSync(closeReq)
-
-    let openReq = SKRequestDictionary(sourcekitd: sourcekitd)
-    openReq[keys.request] = requests.editor_open
-    openReq[keys.name] = url.path
-    openReq[keys.sourcetext] = snapshot.text
-    if let settings = newSettings {
-      openReq[keys.compilerargs] = settings.compilerArguments
-    }
+    // Move to the queue into order to interface with sourcekitd.
+    self.queue.async {
+      let keys = self.keys
+
+      // Close and re-open the document internally to inform sourcekitd to
+      // update the settings. At the moment there's no better way to do this.
+      let closeReq = SKRequestDictionary(sourcekitd: self.sourcekitd)
+      closeReq[keys.request] = self.requests.editor_close
+      closeReq[keys.name] = url.path
+      _ = self.sourcekitd.sendSync(closeReq)
+
+      let openReq = SKRequestDictionary(sourcekitd: self.sourcekitd)
+      openReq[keys.request] = self.requests.editor_open
+      openReq[keys.name] = url.path
+      openReq[keys.sourcetext] = snapshot.text
+      if let settings = newSettings {
+        openReq[keys.compilerargs] = settings.compilerArguments
+      }
 
-    guard let dict = self.sourcekitd.sendSync(openReq).success else {
-      // Already logged failure.
-      return
+      guard let dict = self.sourcekitd.sendSync(openReq).success else {
+        // Already logged failure.
+        return
+      }
+      self.queue_publishDiagnostics(response: dict, for: snapshot)
     }
-    publishDiagnostics(response: dict, for: snapshot)
   }
 
   // MARK: - Text synchronization
 
-  func openDocument(_ note: Notification<DidOpenTextDocument>) {
+  public func openDocument(_ note: DidOpenTextDocument) {
     guard let snapshot = documentManager.open(note) else {
       // Already logged failure.
       return
     }
 
+    let url = snapshot.document.url
+
+    // Cache the `BuildSystem`'s settings interally.
+    let settings = buildSystem.settings(for: url, snapshot.document.language)
+    buildSettingsByFile[url] = settings
```

Comment:
Don't we need to be on `queue` to access this?

Reply:
Currently I have this modified outside of the queue but I can move it onto the queue if you'd like

Reply:
I meant, doesn't it need that to be thread-safe?

Reply:
We could rely on always being called on the `SourceKitServer` queue, but I guess for safety we should move it over to its own queue?

Reply:
Swapped over all calls to its own queue